examples/android/OGVideoProcDroid/src/ogles_gpgpu/examples/ogvideoprocdroid/CameraProcThread.java

```diff
@@ -0,0 +1,404 @@
+    private int newWidth = -1;
+    private int newHeight = -1;
+
+    /**
+      * Flag: Button clicked event received
+      */
+    private boolean modeChanged = false;
+
+    /**
+      * Flag for new mode.
+      * - true: new mode is "input"
+      * - false: new mode is "output"
+      */
+    private boolean newModeInput = false;
+
+    /**
+     * Constructor. Called either from surfaceCreated() or, when the
+     * surface still exists, directly from onResume().
+     */
+    public CameraProcThread(CPUImgProcThread thread, SurfaceHolder surfaceHolder) {
+        imgProcThread = thread;
+        holder = surfaceHolder;
+    }
+
+    /**
+     * Thread run loop. Will run until terminate() is called.
+     */
+    @Override
+    public void run() {
+        Log.i(CamActivity.TAG, "starting thread CameraProcThread");
+
+        // Open the camera device.
+        Camera cam = startCam();
+
+        // init ogles_gpgpu: use platform optimizations, do NOT init
+        // EGL (already done before) and create a render display.
+        Log.i(CamActivity.TAG, "initializing ogles_gpgpu");
+        ogWrapper.init(true, false, true);
+
+        // create window surface and make it the current window
+        Log.i(CamActivity.TAG, "creating window surface");
+        windowSurface = new WindowSurface(eglCore, holder.getSurface(), false);
+        windowSurface.makeCurrent();
+
+        // create camera frame texture and this Activity as callback for
+        // new camera frames
+        Log.i(CamActivity.TAG, "creating camera frame texture");
+        camTextureId = createGLTexture();
+        camTexture = new SurfaceTexture(camTextureId);
+        camTexture.setOnFrameAvailableListener(this);
+
+        Log.i(CamActivity.TAG, "created camera frame texture with id " + camTextureId);
+
+        // set "camTexture" as the Camera object's preview texture
+        try {
+            cam.setPreviewTexture(camTexture);
+        } catch (IOException e) {
+            e.printStackTrace();
+            Log.e(CamActivity.TAG, "setPreviewTexture failed");
+        }
+
+        // start the camera preview output
+        Log.i(CamActivity.TAG, "starting camera preview");
+        cam.startPreview();
+
+        try {
+            mainLoop();
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+            Log.e(CamActivity.TAG, "CameraProcThread mainLoop");
+        }
+
+        stopCam(cam);
+
+        Log.i(CamActivity.TAG, "releasing ogles_gpgpu");
+        ogWrapper.cleanup();
+
+        // Reset EGL state.
+        eglCore.makeNothingCurrent();
+        eglCore.release();
+
+        // release OpenGL resources and the WindowSurface
+        releaseGL();
+
+        Log.i(CamActivity.TAG, "stopped thread CameraProcThread");
+    }
+
+    private void mainLoop() throws Exception {
+        while (running) {
+            boolean frameAvailable = false;
+            synchronized (lock) {
+                if (isSurfaceChanged) {
```

Comment:
This is kind of processing of different events from main UI thread. I think [Handler](http://developer.android.com/reference/android/os/Handler.html) might be useful here. But I'm not so skilled Java user to figure out how it can be implemented yet.